x/config/spec.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	if spec.BuildTransport == nil {
+		return nil, errors.New("BuildTransport is required")
+	}
+
+	var errors []error
+
+	// Helper to chain together the compile calls
+	appendError := func(cs *configSpec, err error) *configSpec {
+		if err != nil {
+			errors = append(errors, err)
+		}
+		return cs
+	}
+
+	out.Transport = appendError(compileTransportConfig(spec.BuildTransport))
+	if spec.BuildInbound != nil {
+		out.Inbound = appendError(compileInboundConfig(spec.BuildInbound))
+	}
+	if spec.BuildUnaryOutbound != nil {
+		out.UnaryOutbound = appendError(compileUnaryOutboundConfig(spec.BuildUnaryOutbound))
+	}
+	if spec.BuildOnewayOutbound != nil {
+		out.OnewayOutbound = appendError(compileOnewayOutboundConfig(spec.BuildOnewayOutbound))
+	}
+	return &out, errs.CombineErrors(errors...)
+}
+
+func (s *compiledTransportSpec) SupportsUnaryOutbound() bool {
+	return s.UnaryOutbound != nil
+}
+
+func (s *compiledTransportSpec) SupportsOnewayOutbound() bool {
+	return s.OnewayOutbound != nil
+}
+
+var (
+	_typeOfError          = reflect.TypeOf((*error)(nil)).Elem()
+	_typeOfTransport      = reflect.TypeOf((*transport.Transport)(nil)).Elem()
+	_typeOfInbound        = reflect.TypeOf((*transport.Inbound)(nil)).Elem()
+	_typeOfUnaryOutbound  = reflect.TypeOf((*transport.UnaryOutbound)(nil)).Elem()
+	_typeOfOnewayOutbound = reflect.TypeOf((*transport.OnewayOutbound)(nil)).Elem()
+)
+
+// Validated representation of a configuration function specified by the user.
+type configSpec struct {
+	inputType reflect.Type  // type of config object expected by the function
+	builder   reflect.Value // function to call
+}
+
+// Decode the configuration for this type from the data map.
+func (cs *configSpec) Decode(attrs attributeMap) (*configuredValue, error) {
+	result := reflect.New(cs.inputType)
+	if err := attrs.Decode(result.Interface()); err != nil {
+		return nil, fmt.Errorf("failed to decode %v: %v", cs.inputType, err)
+	}
+	return &configuredValue{builder: cs.builder, data: result.Elem()}, nil
+}
+
+// A single parsed configuration.
+type configuredValue struct {
+	// Decoded configuration data
+	data reflect.Value
+
+	// A function that accepts Config as its first argument and returns a
+	// result and an error.
+	//
+	// 	func(Config, ...) (O, error)
+	//
+	// Build(...) will call this function and interpret the result.
+	builder reflect.Value
+}
+
+// Build the object configured by this value. The arguments are passed to the
+// build function with the underlying configuration as the first parameter.
+//
+// Arguments may be reflect.Value objects or any other type.
+func (cv *configuredValue) Build(args ...interface{}) (interface{}, error) {
+	callArgs := make([]reflect.Value, len(args)+1)
+	callArgs[0] = cv.data
+
+	for i, v := range args {
+		if value, ok := v.(reflect.Value); ok {
+			callArgs[i+1] = value
+		} else {
+			callArgs[i+1] = reflect.ValueOf(v)
+		}
+	}
+
+	result := cv.builder.Call(callArgs)
+	err, _ := result[1].Interface().(error)
```

Comment:
The cast will fail only if `err` is nil since we've already verified that the function signature matches.